Run tests on HHVM

photodude opened the pull request:
.

photodude committed:
Run tests on HHVM

zero-24 commented:
@photodude do you know why it fails on
`cp: cannot create regular file `/home/travis/.phpenv/versions/hhvm/etc/conf.d': No such file or directory`

https://travis-ci.org/joomla/joomla-cms/jobs/60998558

mbabker commented:
HHVM isn't native PHP, so stuff like `phpenv config-add` in our config scripts fail.  Basically we need to tweak the `.travis.yml` scripts to check for HHVM in extra places (probably the same ones as we are for PHP 7 at a minimum).

photodude committed:
Enable additional PHP extensions for hhvm

- Match php 7 "Enable additional PHP extensions" and scripts for hhvm
- Reorder the matrix to match the order that php versions are called

photodude committed:
Create travis.sh

adapted from information found at https://github.com/travis-ci/travis-ci/issues/2523

photodude committed:
Update .travis.yml

use travis.sh for config

photodude committed:
modify travis.sh permissions

photodude committed:
Update .travis.yml

photodude commented:
@zero-24 I'm not sure about that failure. I've tried tweaking somethings as @mbabker suggested. 

I'm now seeing a test failure `Fatal error: request has exceeded memory limit in /home/travis/build/joomla/joomla-cms/libraries/vendor/phpunit/php-timer/PHP/Timer.php on line 138`

Seems there is a lot to do in relation to our testing with PHP7 and HHVM. These php builds will likely be allowed failures for some time.

zero-24 commented on .travis.yml:
```diff
@@ -19,23 +21,28 @@ before_script:
+  - ./build/travis/travis.sh
+#  - sh -c "if [ '$TRAVIS_PHP_VERSION' != '7.0' ] || [ '$TRAVIS_PHP_VERSION' != 'hhvm' ]; then phpenv config-add build/travis/phpenv/memcached.ini; fi"
+#  - sh -c "if [ '$TRAVIS_PHP_VERSION' != '7.0' ] || [ '$TRAVIS_PHP_VERSION' != 'hhvm' ]; then phpenv config-add build/travis/phpenv/apc-$TRAVIS_PHP_VERSION.ini; fi"
+#  - sh -c "if [ '$TRAVIS_PHP_VERSION' != '7.0' ] || [ '$TRAVIS_PHP_VERSION' != 'hhvm' ]; then phpenv config-add build/travis/phpenv/redis.ini; fi"
```
@photodude as we have now the `travis.sh` i guess we can remove this lines instead of comment it?

photodude replied:
I agree I think those comments can now be removed.

zero-24 commented on build/travis/travis.sh:
```diff
@@ -0,0 +1,16 @@
+    phpenv config-add build/travis/phpenv/memcached.ini
+    phpenv config-add build/travis/phpenv/apc-$VERSION.ini
+    phpenv config-add build/travis/phpenv/redis.ini
+#    echo "extension = memcache.so"  >> $PHPINI
```
@photodude same applys here do we need the commented code or was it only "testing" code that can now removed instead of be commented?

photodude replied:
It was testing initially since we needed something slightly different from the example I used.
Seems we might need to use  `echo "extension = memcache.so"  >> $PHPINI` for PHP 7 since the [documentation](http://docs.travis-ci.com/user/languages/php/#Preinstalled-PHP-extensions) suggests the standard Travis CI PECL extensions are not installed on PHP7. Waiting to see if diff a02f939 works a little better.

mbabker commented:
Maybe we could find someone from the HHVM team to give suggestions on the Travis build config?  FWIW they are running the tests on their platform (see https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/frameworks.yaml#L246) and get past this test (http://graph.facebook.com/hhvm_oss_tests has some raw data from the test runs), only thing is it seems the tests are on the CMS from its state last August (https://github.com/joomla/joomla-cms/tree/7689d922ca373989e7328f510e94aa2fc541e272 specifically).

photodude committed:
Remove commented code replaced by travis.sh

photodude committed:
Changes for php 7

photodude committed:
switch else if for elif

photodude committed:
Create apc-7.0.ini

photodude committed:
add APC to PHP 7

photodude committed:
update composer

- should we be using `- composer self-update` or `- composer update` assuming `- composer update`

photodude committed:
composer self-update seems to be preferred method. 

- test "Set up Apache script" (assuming it will need a `- chmod ugo+x build/travis/php-apache.sh` but testing without)

photodude committed:
Remove php_codesniffer and cache_lite

- moved php_codesniffer and cache_lite to composer.json

photodude committed:
Update .travis.yml

```
The command "sh -c "if [ '$TRAVIS_PHP_VERSION' != '7.0' ] || [ '$TRAVIS_PHP_VERSION' != 'hhvm' ]; then libraries/vendor/bin/phpcs --report=full --extensions=php -p --standard=build/phpcs/Joomla .; fi"" exited with 127.
```
Try without this line

photodude committed:
Add support for PostgreSQL to HHVM

- See reference [PostgreSQL for HHVM releases](https://github.com/PocketRent/hhvm-pgsql/tree/releases)

photodude committed:
Update .travis.yml

photodude committed:
PDO driver for PostgreSQL currently unsupported by HHVM

- also try phpcs report settings without if statemement

photodude committed:
PDO driver for PostgreSQL currently unsupported by HHVM

- requires a [third party dependency](https://github.com/PocketRent/hhvm-pgsql)

photodude committed:
Add php_codesniffer and pear/cache_lite

- add squizlabs/php_codesniffer and pear/cache_lite to `composer.json` so they can be removed from `.travis.yml`
- using subversion `*` for the required projects to reduce need to update this file 
-attempt phpunit 4.6

photodude committed:
Update .travis.yml

- It looks like composer might be installing phpcs to a different location

photodude committed:
Update .travis.yml

photodude committed:
Some suggestions 

- PHP Mess Detector
- PHP Docblock Checker

photodude committed:
Update .travis.yml

photodude committed:
Missing comma

photodude committed:
Update .travis.yml

photodude committed:
Update composer.json

photodude committed:
Update .travis.yml

photodude committed:
Use apcu in PHP 5.5+

photodude committed:
Use apcu for php5.5+

photodude committed:
Update composer.json

Try required rather than required-dev
Try adding custom path if that doesn't work.
```     "extra": {
        "installer-paths": {
            "libraries/vendor/bin/phpcs/": ["shama/ftp", "squizlabs/php_codesniffer"]
        }
    }
``

photodude committed:
Try installing apcu

photodude committed:
moving php_codesniffer and pear/cache_lite back to travis.yml

photodude committed:
moving php_codesniffer and pear/cache_lite back to travis.yml

photodude committed:
Update travis.sh

photodude committed:
Update .travis.yml

photodude committed:
Update composer.json

photodude committed:
Update .travis.yml

photodude committed:
Try installing apcu via channel

photodude committed:
pecl channel-update

"pecl.php.net" has updated its protocols

photodude committed:
Update .travis.yml

photodude committed:
Create install-apcu.sh

photodude committed:
add install-apcu.sh

photodude committed:
Delete install-apcu.sh

photodude committed:
better apcu installation commmand

photodude committed:
Module 'apcu' already loaded in Unknown on line 0?

photodude committed:
Update .travis.yml

The `pecl' command does not exist in these PHP versions: 7, hhvm

photodude committed:
Update apc-5.6.ini

photodude committed:
Update apc-5.6.ini

photodude committed:
try nesting the apcu install for 5.5+

photodude committed:
GitHub's rate limits on their API was hit

need to set up an OAuth token so Composer can go ahead with its work https://github.com/composer/composer/blob/master/doc/articles/troubleshooting.md#api-rate-limit-and-two-factor-authentication

photodude committed:
what line was that...

photodude committed:
Update travis.sh

photodude committed:
Update travis.sh

photodude committed:
Try installing apcu via channel for 5.5+ and not hhvm

photodude committed:
Update .travis.yml

photodude committed:
separate  pecl commands excluded from hhvm

photodude commented:
@mbabker I've submitted a PR https://github.com/facebook/hhvm/pull/5249 for hhvm to update to the current 3.4.1 release.

photodude committed:
phpenv config-add does the same thing as echo to php.ini

photodude committed:
Update .travis.yml

photodude commented:
Interesting, Travis is now saying this PR is "missing config" on each commit and no longer is running tests. Seems to be working for the rest of the repo.

mbabker commented on .travis.yml:
```diff
@@ -6,36 +6,45 @@ php:
+- sh -c "if [ '$TRAVIS_PHP_VERSION' != 'hhvm' ]; then echo -e "yes\nno\n" | pecl -d preferred_state=beta install apcu; fi"
+  #- composer update
+  #- composer self-update
+    # Install PHPCS to validate code standards
```
Since YAML is picky on spacing, try to remove the extra spaces on this line and see what happens.

photodude committed:
remove whitespace

- "YAML is picky on spacing, try to remove the extra spaces on this line and see what happens."

photodude committed:
YAML is picky on spacing

YAML is picky on spacing, try indenting recently added lines and see what happens.

photodude commented:
@mbabker Thanks, I didn't know YAML was so picky on spacing. Looks like I needed to remove the referenced whitespace and I needed to indent some lines.

photodude committed:
Try with install method

mbabker commented on .travis.yml:
```diff
@@ -6,36 +6,45 @@ php:
+install:
+  - sh -c "if [ '$TRAVIS_PHP_VERSION' != 'hhvm' ]; then pecl channel-update pecl.php.net; fi"
+  - sh -c "if [ '$TRAVIS_PHP_VERSION' != 'hhvm' ]; then echo -e "yes\nno\n" | pecl -d preferred_state=beta install apcu; fi"
+  - travis_retry composer self-update && composer update
```
We can't have the composer update part of this script run otherwise it updates the Composer dependencies to versions not locked in via the lock file.  We shouldn't need to do the composer self-update either as we aren't really doing anything special with Composer where we'd need a version later than what Travis supplies.

photodude replied:
Sure thing, would you like it taken out, or just commented like it was previously? From the history I only saw that `composer update` was commented due to an issue with gihub authentication.

mbabker replied:
I think we can just take it out completely at this point.

mbabker commented on composer.json:
```diff
@@ -18,11 +18,15 @@
         "ircmaxell/password-compat": "1.*",
         "leafo/lessphp": "0.3.9",
-        "phpmailer/phpmailer": "5.2.9",
+        "phpmailer/phpmailer": "5.2.*",
```
Please leave this as `5.2.9` for the time being.

photodude committed:
Set phpmailer to 5.2.9

phpmailer greater than 5.2.9 will need to be considered and tested for 3.5 or 4

photodude committed:
Update .travis.yml

- "We can't have the composer update part of this script run otherwise it updates the Composer dependencies to versions not locked in via the lock file. We shouldn't need to do the composer self-update either as we aren't really doing anything special with Composer where we'd need a version later than what Travis supplies."
- Additionally composer has an issue with authentication `Could not authenticate against github.com` resulting in intermittent build failures. we would need to set up an OAuth token so Composer can go ahead with its work https://github.com/composer/composer/blob/master/doc/articles/troubleshooting.md#api-rate-limit-and-two-factor-authentication

photodude committed:
more tweaks

- test with `phpenv rehash` since we added some stuff

photodude committed:
Joomla testing scripts doesn't like separating the databases

photodude committed:
try a different apcu install command

photodude committed:
attempt checks for apcu being installed

photodude committed:
check for APCU installed failed...

photodude committed:
try moving the check for ge php 5.5

photodude committed:
Update .travis.yml

photodude committed:
No need to enable apcu since we are installing it.

photodude committed:
No need to enable apcu since we are installing it.

photodude committed:
Try installing a few dependencies for php 7 and hhvm

photodude committed:
Create apc-hhvm.ini

photodude committed:
Comments start with ';', as in php.ini

photodude committed:
Comments start with ';', as in php.ini

photodude committed:
Comments start with ';', as in php.ini

photodude committed:
remove sudo apt-get install -y hhvm-dbg since it's erroring

photodude committed:
To correctly install to install redis it may require the redis-server

photodude committed:
Update travis.sh

photodude committed:
Try installing pear for php 7

photodude committed:
redis-server not needed for php7

photodude committed:
memcached does not need to be installed for php 7

photodude committed:
Update .travis.yml

photodude committed:
PHP 7 needs zend-framework

photodude committed:
Use = rather than == in the test expression

photodude committed:
HHVM doesn't need to install memcached. libmemcached, or redis-server

photodude committed:
Try simplifying the apcu install

photodude committed:
memory_limit = -1 didn't help with phpunit/php-timer/PHP/Timer.php exceeded memory limit issue

photodude committed:
modify APCu install

photodude committed:
Retry the earlier APCu install method

photodude committed:
try modifying the pear install

photodude commented:
@mbabker Looking over the information at https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/frameworks.yaml#L246

it seems the HHVM team is indicating something about tests intermittently hanging...
https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/frameworks.yaml#L259

photodude committed:
php 7 pear install is having issues

- exclude php 7 from pecl installs

mbabker commented:
I saw that, no idea what the deal is.  At least it's all isolated to a single package so you'd think there would be something common in there.

photodude committed:
pecl install of apcu isn't needed?

photodude committed:
memory_limit = -1 didn't help with phpunit/php-timer/PHP/Timer.php exceeded memory limit issue

- tweak php 7 to see if we can eliminate the following errors
```
PHP Warning:  PHP Startup: Unable to load dynamic library '/home/travis/.phpenv/versions/nightly/lib/php/extensions/no-debug-zts-20141001/apcu.so' - /home/travis/.phpenv/versions/nightly/lib/php/extensions/no-debug-zts-20141001/apcu.so: cannot open shared object file: No such file or directory in Unknown on line 0

PHP Warning:  PHP Startup: Unable to load dynamic library '/home/travis/.phpenv/versions/nightly/lib/php/extensions/no-debug-zts-20141001/memcached.so' - /home/travis/.phpenv/versions/nightly/lib/php/extensions/no-debug-zts-20141001/memcached.so: cannot open shared object file: No such file or directory in Unknown on line 0

PHP Warning:  PHP Startup: Unable to load dynamic library '/home/travis/.phpenv/versions/nightly/lib/php/extensions/no-debug-zts-20141001/redis.so' - /home/travis/.phpenv/versions/nightly/lib/php/extensions/no-debug-zts-20141001/redis.so: cannot open shared object file: No such file or directory in Unknown on line 0
```

photodude committed:
Check what apt-get sources are listed for php7 and hhvm

photodude committed:
Try adding the apt-get universe repository

photodude committed:
Unfortunetly 'universe' repository is invalid on php 7 and hhvm

photodude committed:
apt-get php5-apcu and php5-redis not found, comment out for now.

photodude committed:
Can we install software-properties-common to find php5-apcu and php5-redis?

photodude committed:
Try to add a repository

photodude committed:
apcu and redis still didn't install for php 7 and hhvm

photodude committed:
Try adding xdebug

photodude committed:
Try pecl install for apcu

photodude committed:
Remove pecl from hhvm

The `pecl' command only exists in these PHP versions:
  5.2
  5.2.17
  5.3
  5.3.29
  5.3.3
  5.4
  5.4.37
  5.5
  5.5.21
  5.5.9
  5.6
  5.6.5

photodude committed:
install php5-mysql

photodude committed:
Most PHP extensions are not installed on php7 and HHVM builds

- Get a list of packages we can use since there are no good references I can find

photodude committed:
apt-cache pkgnames crashes

photodude committed:
Update .travis.yml

- change the ordering of the items.

photodude committed:
change the php 7 installs

- php5-mysql is installed
- php-apc is available not php5-apcu?
-

photodude committed:
More php 7 and hhvm modifications

photodude committed:
Add suggested packages

photodude committed:
Try changing the hhvm phpunit location

photodude committed:
Try bumping phpunit to 4.6.6

- **note:** there are inconsistencies in the phpunit versions listed in the lock file. references locked to 4.1.6 up to 4.4.* or 4.* or 4.1.* or ~4.3 This should be cleaned up following the phpunit 4.6.6 check

photodude committed:
Try bumping phpunit to 4.6.6

photodude committed:
Update .travis.yml

photodude commented:
@mbabker It would be nice if the issue is a single package. I fear it's not.

Looking through the various builds I find a few issues.
Something is throwing an `Invalid configuration directive` issue

There is this memory limit issue
```
Fatal error: request has exceeded memory limit in /home/travis/build/joomla/joomla-cms/libraries/vendor/phpunit/php-timer/PHP/Timer.php on line 138
/home/travis/build.sh: line 41:  5463 Segmentation fault      phpunit --configuration travisci-phpunit.xml
```
and this abstract method issue 
```
Fatal error: Class Mock_JDatabaseDriver_afa00a82 contains abstract method (isSupported) and must therefore be declared abstract or implement the remaining methods in /home/travis/build/joomla/joomla-cms/libraries/vendor/phpunit/phpunit-mock-objects/src/Framework/MockObject/Generator.php(301) : eval()'d code on line 1
Core dumped: Segmentation fault
Stack trace in /tmp/stacktrace.6888.log
/home/travis/build.sh: line 41:  6888 Segmentation fault      phpunit --configuration travisci-phpunit.xml
```

I've adjusted a few things trying to see if these errors might go away, with no success so far.

Looking at the [HHVM test](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect) results from Facebook there are some failures and errors. 
- [JArrayHelperTest::testSortObjects with data set "by casesensitive string descending"](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L1021)
- [JArrayHelperTest::testSortObjects with data set "by string ascending"](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L1037)
- [JCategoryNodeTest::testSetParent](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L1385)
- [JDocumentRendererMessageTest::testRender](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L3001)
- [JFactoryTest::testGetDocument](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L3925)
- [JHtmlBatchTest::testAccess](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L6835)
- [JHtmlBatchTest::testItem](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L6837)
- [JHtmlBatchTest::testLanguage](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L6839)
- [JHtmlBatchTest::testUser](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L6841)
- [JHtmlListTest::testPositions[(https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L7025)
- [JHtmlListTest::testUsers](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L7027)
- [JHtmlSelectTest::testRadiolist with data set #0](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L7109)
- [JHtmlSelectTest::testRadiolist with data set #1](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L7111)
- [JHtmlTest::testScript](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L7231)
- [JHtmlTest::testStylesheet](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L7235)
- [JLoaderTest::testSetupWithoutPrefixes](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L7991)
- [JLogEntryTest::testDefaultValues](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L7997)
- [JOAuth2ClientTest::testAuth](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L8539)
- [JPaginationTest::testBuildDataObject with data set #0](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L8803)
- [JPaginationTest::testGetData with data set #0](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L8825)
- [JPaginationTest::testGetPagesLinks with data set #0](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L8833)
- [JPaginationTest::testItemActive with data set #0](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L8837)
- [JViewLegacyTest::test__Construct](https://github.com/facebook/hhvm/blob/master/hphp/test/frameworks/results/joomlacms.expect#L10993)

Maybe this list will give us some direction on what needs to be adjusted or fixed.

photodude committed:
Add some comments and use [[ as best practice

photodude committed:
Remove sh - c

photodude committed:
Use libraries/vendor/bin/phpunit rather than phpunit

- We are possibly not using the installed phpunit version which is in the "composer.json" and "composer.lock" files but the Travis version

photodude committed:
Fix Typo, wrong bracket: use ] not }

photodude commented:
@zero-24 should this PR get a "Unit/System Tests" label?

zero-24 commented:
done thanks @photodude

photodude commented:
@zero-24 You're welcome, and Thank you.

wilsonge commented:
FWIW with the test sort objects failures - we had a PR for that a LONG time ago - https://github.com/joomla/joomla-cms/pull/3200 but travis didn't like it (although it ran locally). Since we cleaned up a lot of our unit tests since then it may fix itself

photodude commented:
The HHVM team failed the Joomla 3.4.1 testing update.

Summary: Does not appear to be tested. refs facebook/hhvm#5249 - brakes
run

```
ERROR: Could not disable
/data/fio/hhvm/hphp/test/frameworks/framework_downloads/joomla-cms/tests/unit/suites/libraries/cms/form/field/JFormFieldCaptchaTest.php
in joomlacms!
```

photodude commented:
Closing as replaced by #9207
